src/Microsoft.VisualStudio.Threading.Tests/AsyncSemaphoreTests.cs

```diff
@@ -161,6 +250,217 @@ public void TwoResourceSemaphore()
         }
 
         [Fact]
+        public async Task CurrentCount()
+        {
+            const int initialCapacity = 3;
+            var sem = new AsyncSemaphore(initialCapacity);
+            Assert.Equal(initialCapacity, sem.CurrentCount);
+
+            var releasers = new AsyncSemaphore.Releaser[initialCapacity];
+            for (int i = 0; i < initialCapacity; i++)
+            {
+                releasers[i] = await sem.EnterAsync();
+                Assert.Equal(initialCapacity - (i + 1), sem.CurrentCount);
+            }
+
+            // After requesting another beyond its capacity, it should still report 0.
+            var extraReleaser = sem.EnterAsync();
+            Assert.Equal(0, sem.CurrentCount);
+
+            for (int i = 0; i < initialCapacity; i++)
+            {
+                releasers[i].Dispose();
+                Assert.Equal(i, sem.CurrentCount);
+            }
+
+            extraReleaser.Result.Dispose();
+            Assert.Equal(initialCapacity, sem.CurrentCount);
+        }
+
+        [Fact]
+        public async Task TooManyReleases_SameStruct()
+        {
+            var releaser = await this.lck.EnterAsync();
+            releaser.Dispose();
+            releaser.Dispose();
+            Assert.Equal(2, this.lck.CurrentCount);
+        }
+
+        [Fact]
+        public async Task TooManyReleases_CopyOfStruct_OverInitialCount()
+        {
+            var releaser = await this.lck.EnterAsync();
+            var releaserCopy = releaser;
+
+            releaser.Dispose();
+            Assert.Equal(1, this.lck.CurrentCount);
+            releaserCopy.Dispose();
+            Assert.Equal(2, this.lck.CurrentCount);
+        }
+
+        [Fact]
+        public async Task TooManyReleases_CopyOfStruct()
+        {
+            var sem = new AsyncSemaphore(2);
+            var releaser1 = await sem.EnterAsync();
+            var releaser2 = await sem.EnterAsync();
+
+            // Assigning the releaser struct to another local variable copies it.
+            var releaser2Copy = releaser2;
+
+            // Dispose of each copy of the releaser.
+            // The double-release is undetectable. The semaphore should be back at capacity 2.
+            releaser2.Dispose();
+            Assert.Equal(1, sem.CurrentCount);
+            releaser2Copy.Dispose();
+            Assert.Equal(2, sem.CurrentCount);
+
+            releaser1.Dispose();
+            Assert.Equal(3, sem.CurrentCount);
```

Comment:
Ya. I thought that was pretty bizarre when I learned `SemaphoreSlim` offered that ability as well. And as `AsyncSemaphore` is supposed to be roughly a more friendly version of `SemaphoreSlim`, rather than prevent this ability (which would have been difficult anyway) I decided to allow it.